tools/codesigningtool/codesigningtool.py

```diff
@@ -120,16 +123,100 @@ def _certificate_fingerprint(identity):
   return fingerprint
 
+def _certificate_common_name(cert):
+  _, subject, _ = execute.execute_and_filter_output([
+    "openssl",
+    "x509",
+    "-noout",
+    "-inform",
+    "DER",
+    "-subject"
+  ], inputstr=cert, raise_on_failure=True)
+  subject = subject.strip().split('/')
+  cert_cn = [f for f in subject if "CN=" in f][0]
+  cert_cn = cert_cn.replace("CN=", "")
+  return cert_cn
 
 def _get_identities_from_provisioning_profile(mpf):
   """Iterates through all the identities in a provisioning profile, lazily."""
   for identity in mpf["DeveloperCertificates"]:
-    if not isinstance(identity, bytes):
+    cert = _certificate_data(identity)
+    yield _certificate_fingerprint(cert)
+
+def _certificate_data(cert):
+  if not isinstance(cert, bytes):
       # Old versions of plistlib return the deprecated plistlib.Data type
       # instead of bytes.
-      identity = identity.data
-    yield _certificate_fingerprint(identity)
+      cert = cert.data
+
+  return cert
 
+def _find_smart_card_identities(identity=None):
+  """Finds smartcard identitites on the current system."""
+  ids = []
+  # Use 'system_profiler' to export XML
+  # containing smartcards keychain data
+  _, xml, _ = execute.execute_and_filter_output([
+      "/usr/sbin/system_profiler",
+      "SPSmartCardsDataType",
+      "-xml"
+  ], raise_on_failure=True)
+  xml = ET.fromstring(xml)
+  # Find 'AVAIL_SMARTCARDS_TOKEN' node and move up one level
+  # to find the 'array' of available tokens, each one is a 'string' node
+  available_tokens_xml = xml.findall('.//*/string[.="AVAIL_SMARTCARDS_TOKEN"]../array')[0]
+  available_tokens_xml = available_tokens_xml.findall('.//*/string')
+  tokens = [t.text for t in available_tokens_xml]
+  # Find 'AVAIL_SMARTCARDS_KEYCHAIN' node and move up one level
+  # to find the 'array' of available keychain items
+  smartcards_keychain = xml.findall('.//*/string[.="AVAIL_SMARTCARDS_KEYCHAIN"]../array')[0]
+  # For each 'token' decode the certificate and for each non-expired cert:
+  #
+  # 1. Check if 'identity' was provided and if it matches a 'CN', in that case stop the loop
+  #    and return the respective `SHA1`
+  # 2. Otherwise append each `SHA1` found to 'ids' to be returned at the end
+  #
+  # ps: note that if 'identity' is provided and it does not match any existing item in the
+  # smartcard keychain 'ids' will be empty, so this function's behaviour is consistent with
+  # '_find_codesign_identities' where it's being called
+  for token in tokens:
+    # Used to find certificate info
+    cert_start = "-----BEGIN CERTIFICATE-----"
+    # All 'string' nodes at the same level of 'token'
+    all_string_nodes = smartcards_keychain.findall('.//*/string[.="{}"]../string'.format(token))
+    # Find only the 'string' nodes that hold certificate data and are not empty
+    cert_data = [s.text for s in all_string_nodes if cert_start in s.text and s.text.strip() != ""]
+
+    for data in cert_data:
+      data_split = data.split('\n')
```

Comment:
Nit: could also do `data.splitlines()` here or below

Reply:
Not necessary anymore since I'm using a regex to get the cert